src/Symfony/Component/PropertyAccess/PropertyAccessor.php

```diff
@@ -253,33 +256,31 @@ private function &readPropertiesUntil(&$objectOrArray, PropertyPathInterface $pr
     /**
      * Reads a key from an array-like structure.
      *
-     * @param \ArrayAccess|array $array The array or \ArrayAccess object to read from
-     * @param string|int         $index The key to read
+     * @param array      $zval  The array containing the array or \ArrayAccess object to read from
+     * @param string|int $index The key to read
      *
-     * @return mixed The value of the key
+     * @return array The array containing the value of the key
      *
      * @throws NoSuchPropertyException If the array does not implement \ArrayAccess or it is not an array
      */
-    private function &readIndex(&$array, $index)
+    private function readIndex($zval, $index)
     {
-        if (!$array instanceof \ArrayAccess && !is_array($array)) {
-            throw new NoSuchPropertyException(sprintf('Index "%s" cannot be read from object of type "%s" because it doesn\'t implement \ArrayAccess', $index, get_class($array)));
+        if (!$zval[self::VALUE] instanceof \ArrayAccess && !is_array($zval[self::VALUE])) {
+            throw new NoSuchPropertyException(sprintf('Index "%s" cannot be read from object of type "%s" because it doesn\'t implement \ArrayAccess', $index, get_class($zval[self::VALUE])));
         }
 
-        // Use an array instead of an object since performance is very crucial here
-        $result = array(
-            self::VALUE => null,
-            self::IS_REF => false,
-        );
+        $result = self::$resultProto;
 
-        if (isset($array[$index])) {
-            if (is_array($array)) {
-                $result[self::VALUE] = &$array[$index];
-                $result[self::IS_REF] = true;
-            } else {
-                $result[self::VALUE] = $array[$index];
+        if (isset($zval[self::VALUE][$index])) {
+            $result[self::VALUE] = $zval[self::VALUE][$index];
+
+            if (!isset($zval[self::REF])) {
+                // Save creating references when doing read-only lookups
+            } elseif (is_array($zval[self::VALUE])) {
+                $result[self::REF] = &$zval[self::REF][$index];
+            } elseif (is_object($result[self::VALUE])) {
                 // Objects are always passed around by reference
```

Comment:
this comment is not accurate (but we don't need a reference when dealing with objects). I suggest updating it to avoid confusing people reading the code